File: 0121_best_time_to_buy_and_sell_stocks/prefix_minimum_array/python/solution.py
def max_profit(prices: list[int]) -> int:
    prefix_min = [0] * len(prices)
    prefix_min[0] = prices[0]

    for i in range(1, len(prices)):
        prefix_min[i] = min(prefix_min[i - 1], prices[i])

    best_profit = 0
    for price, buy_price in zip(prices, prefix_min):
        best_profit = max(best_profit, price - buy_price)

    return best_profit

File: 0121_best_time_to_buy_and_sell_stocks/prefix_minimum_array/go/solution.go
package best_time_to_buy_and_sell_stock

func maxProfit(prices []int) int {
	prefixMin := make([]int, len(prices))
	prefixMin[0] = prices[0]

	for i := 1; i < len(prices); i++ {
		if prices[i] < prefixMin[i-1] {
			prefixMin[i] = prices[i]
		} else {
			prefixMin[i] = prefixMin[i-1]
		}
	}

	bestProfit := 0
	for i, price := range prices {
		if profit := price - prefixMin[i]; profit > bestProfit {
			bestProfit = profit
		}
	}

	return bestProfit
}
